dipy/reconst/mapmri.py

```diff
@@ -0,0 +1,695 @@
+               and estimation for one-dimensional q-space magnetic resonance
+               1D-SHORE)", eapoc Intl Soc Mag Reson Med, vol. 16, p. 35., 2008.
+
+        .. [3] Ozarslan E. et. al, "Simple harmonic oscillator based reconstruction
+               and estimation for three-dimensional q-space mri", ISMRM 2009.
+
+        Examples
+        --------
+        In this example, where the data, gradient table and sphere tessellation
+        used for reconstruction are provided, we model the diffusion signal
+        with respect to the MAPMRI model and compute the real and analytical
+        ODF.
+
+        from dipy.data import get_data,get_sphere
+        sphere = get_sphere('symmetric724')
+        fimg, fbvals, fbvecs = get_data('ISBI_testing_2shells_table')
+        bvals, bvecs = read_bvals_bvecs(fbvals, fbvecs)
+        gtab = gradient_table(bvals, bvecs)
+        from dipy.sims.voxel import SticksAndBall
+        data, golden_directions = SticksAndBall(gtab, d=0.0015,
+                                                S0=1, angles=[(0, 0), (90, 0)],
+                                                fractions=[50, 50], snr=None)
+        from dipy.reconst.mapmri import MapmriModel
+        radial_order = 4
+        map_model = MapmriModel(gtab, radial_order=radial_order)
+        mapfit = map_model.fit(data)
+        odf= mapfit.odf(sphere)
+        """
+
+        self.bvals = gtab.bvals
+        self.bvecs = gtab.bvecs
+        self.gtab = gtab
+        self.radial_order = radial_order
+        self.lambd = lambd
+        self.eap_cons = eap_cons
+        self.anisotropic_scaling=anisotropic_scaling
+        if (gtab.big_delta is None) or (gtab.small_delta is None):
+            self.tau = 1 / (4 * np.pi ** 2)
+        else:
+            self.tau = gtab.big_delta - gtab.small_delta / 3.0
+        self.eigenvalue_threshold = eigenvalue_threshold
+        self.bmax_threshold = bmax_threshold
+
+    @multi_voxel_fit
+    def fit(self, data):
+        ind=self.gtab.bvals<=self.bmax_threshold
+        gtab2 = gradient_table(self.gtab.bvals[ind], self.gtab.bvecs[ind,:])
+        tenmodel=dti.TensorModel(gtab2)
+        tenfit=tenmodel.fit(data[...,ind])
+        
+        evals = tenfit.evals
+        R=tenfit.evecs
+
+        ind_evals = np.argsort(evals)[::-1]
+        evals = evals[ind_evals]
+        R = R[ind_evals,:]
+        
+        evals = np.clip(evals,self.eigenvalue_threshold,evals.max())
+        
+        if self.anisotropic_scaling:
+            mu = np.sqrt(evals*2*self.tau)
+            
+        else:
+            mumean=np.sqrt(evals.mean()*2*self.tau)
+            mu=np.array([mumean,mumean,mumean])
+        
+        qvals=np.sqrt(self.gtab.bvals/self.tau) / (2 * np.pi)
+        qvecs=np.dot(self.gtab.bvecs,R)
+        q=qvecs*qvals[:,None]
+
+        M = mapmri_phi_matrix(self.radial_order, mu, q.T)
+        
+        ind_mat = mapmri_index_matrix(self.radial_order)
```

Comment:
There are a few things here that could be moved outside of the fit function. For example `ind_mat` and `tenmodel` don't need to be re-built at each voxel. Instead you could make them in the init and re-use them.

It might make your `fit` method easier to follow because currently it's kind of long. Having only the steps that are specific to individual voxels might make the method easier to understand.